### PR TITLE
Issue #7306: fix RequireThis for method reference

### PR DESCRIPTION
**Pull request by strkkk**
#7306

Report https://strkkk.github.io/checkstyle/7306/index.html

Report for spring-integration https://strkkk.github.io/checkstyle/7306_2/index.html

### CODE REVIEW

**Comment by rnveach** [comment]
Please add regression for projects that reported the issue to make sure they are fixed.
https://github.com/checkstyle/checkstyle/issues/7306#issuecomment-561796516

We could have 2 additional reports to make sure it is fixed.

Normal regression should show violations being removed, which is just the bug identified in this issue.

Extra regression should show the difference in the check with master branch being before your previous issue that introduced the current bug and pr branch being this PR's branch. This regression should show no differences at all (additions or removals). This would signify their projects are clear of issues with this check that we know of.

Does this make sense?

**Comment by romani** [comment]
Please send PRs to add spring-integrations to contribution config.
It would be awesome if you could add such project to "no-error" to wercker.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisValidateOnlyOverlappingFalse.java [hunk 1]
```diff
@@ -478,3 +478,12 @@ private void hashCodeMethod(String param) {
         hashCode();
     }
 }
+class Issue7306 {
+    java.util.List<String> add = new java.util.ArrayList<>();
+
+    void someMethod() {
+        java.util.List<String> test = new java.util.ArrayList<>();
+        test.forEach(this.add::add); // OK
+        test.forEach(add::add); // violation
```
Comment (romani) [inline src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisValidateOnlyOverlappingFalse.java]: Please add test for reference by class name https://strkkk.github.io/checkstyle/7306/Vavr/xref/checkstyle/contribution/checkstyle-tester/repositories/Vavr/vavr/src/main/java/io/vavr/collection/HashArrayMappedTrie.java.html#L473

And https://strkkk.github.io/checkstyle/7306/guava/xref/checkstyle/contribution/checkstyle-tester/repositories/guava/guava-testlib/src/com/google/common/collect/testing/AbstractIteratorTester.java.html#L330 where no `this` and all variables are local to method.

We need to copy all interesting cases from regression report, to less depends on it. In some ideal future, execution on our Inputs should cover all cases/combination of code.

**Comment by romani** [comment]
please rebase, TC is fixed in master.

File: src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisValidateOnlyOverlappingFalse.java (hunk 1 above)
Comment (strkkk) [reply src/test/resources/com/puppycrawl/tools/checkstyle/checks/coding/requirethis/InputRequireThisValidateOnlyOverlappingFalse.java]: done

**Comment by strkkk** [comment]
@romani  PR https://github.com/checkstyle/contribution/pull/419
@rnveach report for spring-integration is added to first post

**Comment by strkkk** [comment]
@romani Build failure is not related with code changed, spellchecker went crazy https://travis-ci.org/checkstyle/checkstyle/jobs/621479097?utm_medium=notification&utm_source=github_status

**Comment by romani** [comment]
Restarted

**Comment by romani** [comment]
What about this ?

>add such project to "no-error" to wercker (separate commit of same PR)

**Comment by strkkk** [comment]
> What about this ?

I have no time and desire to dig into CI things.

In fact, I would suggest to make less builds, because it takes a lot of time, and big number of commits is just to keep it in a working state.

**Comment by romani** [comment]
"no-error" model is the most reliable testing, as it actually trigger conversation with actual author of code, and it is not rare case when author of code is not ok with change.

I will add it myself.

**Pull request merged**